fendermint/actors/f3-cert-manager/src/lib.rs

```diff
@@ -0,0 +1,651 @@
+// Copyright 2021-2024 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use crate::state::State;
+use crate::types::{
+    ConstructorParams, GetCertificateResponse, GetInstanceInfoResponse, PowerEntry,
+    UpdateCertificateParams,
+};
+use fil_actors_runtime::builtin::singletons::SYSTEM_ACTOR_ADDR;
+use fil_actors_runtime::runtime::{ActorCode, Runtime};
+use fil_actors_runtime::{actor_dispatch, actor_error, ActorError};
+use fvm_shared::METHOD_CONSTRUCTOR;
+use num_derive::FromPrimitive;
+
+pub mod state;
+pub mod types;
+
+#[cfg(feature = "fil-actor")]
+fil_actors_runtime::wasm_trampoline!(F3CertManagerActor);
+
+pub const F3_CERT_MANAGER_ACTOR_NAME: &str = "f3_cert_manager";
+
+pub struct F3CertManagerActor;
+
+#[derive(FromPrimitive)]
+#[repr(u64)]
+pub enum Method {
+    Constructor = METHOD_CONSTRUCTOR,
+    UpdateCertificate = frc42_dispatch::method_hash!("UpdateCertificate"),
+    GetCertificate = frc42_dispatch::method_hash!("GetCertificate"),
+    GetInstanceInfo = frc42_dispatch::method_hash!("GetInstanceInfo"),
+    GetGenesisInstanceId = frc42_dispatch::method_hash!("GetGenesisInstanceId"),
+    GetGenesisPowerTable = frc42_dispatch::method_hash!("GetGenesisPowerTable"),
+}
+
+trait F3CertManager {
+    /// Update the latest F3 certificate
+    fn update_certificate(
+        rt: &impl Runtime,
+        params: UpdateCertificateParams,
+    ) -> Result<(), ActorError>;
+
+    /// Get the latest F3 certificate
+    fn get_certificate(rt: &impl Runtime) -> Result<GetCertificateResponse, ActorError>;
+
+    /// Get F3 instance information
+    fn get_instance_info(rt: &impl Runtime) -> Result<GetInstanceInfoResponse, ActorError>;
+
+    /// Get the genesis F3 instance ID
+    fn get_genesis_instance_id(rt: &impl Runtime) -> Result<u64, ActorError>;
+
+    /// Get the genesis power table
+    fn get_genesis_power_table(rt: &impl Runtime) -> Result<Vec<PowerEntry>, ActorError>;
+}
+
+impl F3CertManagerActor {
+    pub fn constructor(rt: &impl Runtime, params: ConstructorParams) -> Result<(), ActorError> {
+        rt.validate_immediate_caller_is(std::iter::once(&SYSTEM_ACTOR_ADDR))?;
+
+        let state = State::new(params.genesis_instance_id, params.genesis_power_table)?;
+
+        rt.create(&state)?;
+        Ok(())
+    }
+}
+
+impl F3CertManager for F3CertManagerActor {
+    fn update_certificate(
```

Comment:
The updated state is not a certificate, but the light client state (finalized chain + power table).

Reply:
What do you mean by finalised chain? Regardless I can store the light client state. Or instance id + power table is also sufficient